packages/core/src/actors.ts

```diff
@@ -128,31 +123,40 @@ export function fromCallback<TEvent extends EventObject>(
                 })
               );
 
-              canceled = true;
+              state.canceled = true;
             }
           );
         }
-        return undefined;
-      } else if (event.type === stopSignalType) {
-        canceled = true;
+        return state;
+      } else if (_event.name === stopSignalType) {
+        state.canceled = true;
 
-        if (isFunction(dispose)) {
-          dispose();
+        if (isFunction(state.dispose)) {
+          state.dispose();
         }
-        return undefined;
+        return state;
       }
 
-      if (isSignal(event)) {
+      if (isSignal(_event.name)) {
         // TODO: unrecognized signal
-        return undefined;
+        return state;
       }
 
       const plainEvent = isSCXMLEvent(event) ? event.data : event;
-      receivers.forEach((receiver) => receiver(plainEvent));
+      if (!isSignal(plainEvent.type)) {
+        state.receivers.forEach((receiver) =>
+          receiver(plainEvent as EventObject)
+        );
+      }
 
-      return undefined;
+      return state;
     },
-    initialState: undefined
+    initialState: {
+      canceled: false,
+      receivers: new Set<(e: EventObject) => void>(),
+      dispose: undefined
+    },
+    getSnapshot: () => undefined
```

Comment:
I think that this function name is misleading and that this has been overloaded too much. It's more like `selectSnapshot` or something.